src/test/java/LoginTests.java

```diff
@@ -3,23 +3,22 @@
 import org.openqa.selenium.chrome.ChromeOptions;
 import org.testng.Assert;
 import org.testng.annotations.Test;
-
 import java.time.Duration;
-
 public class LoginTests extends BaseTest {
     @Test
     public void loginEmptyEmailPassword() {
-
 //      Added ChromeOptions argument below to fix websocket error
         ChromeOptions options = new ChromeOptions();
         options.addArguments("--remote-allow-origins=*");
+        options.addArguments(("--no-sandbox"));
+        options.addArguments("--disable-notifications","--remote-allow-origins=*", "--incognito","--start-maximized");
+        options.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
```

Comment:
**1) GIVEN – WHEN - THEN:**
Each test method is logically split using the GIVEN - WHEN - THEN structure, making the tests easy to read and understand.

- GIVEN describes the starting situation (the user is on the login page).
- WHEN describes the action being taken (the user enters data and clicks the button).
- THEN describes the expected outcome (checking if the user stays on the login page or successfully logs in).


**2) Page Object Model:**
The code follows the Page Object Model (POM) pattern, which improves readability and maintainability.

- The LoginPage and HomePage classes represent the application pages and contain methods to interact with elements on those pages.
- Test methods use these page classes to perform actions and verify results.


**3) Helper Methods in Other Classes:**

- Methods for performing actions and verifying results are defined in the respective page classes (LoginPage and HomePage), following the single responsibility principle and making the code easier to maintain.
- For example, methods like `provideEmail()`, `providePassword()`, and `clickSubmit()` are in the `LoginPage class`, while `getUserAvatarIcon()` is in the `HomePage class`.

**Removed Empty Lines:**
The code has unnecessary empty lines removed, making it more compact and readable.

**Properly Structured Code:**
The code is well-organized, with clear separation of concerns, making it easy to follow and maintain.

_For example_

**LoginTest class following the Page Object Model (POM) and using Page Factory for better code structure and maintainability:**

![image](https://github.com/Test-Pro-Bootcamp/A59-QA-Automation-Framework/assets/104761482/432fecd1-3e3e-4928-b796-e648049f6144)


```
import org.testng.Assert;
import org.testng.annotations.Test;
import page.HomePage;
import pageFactory.LoginPage;

public class LoginTest extends BaseTest {

  @Test
  public void loginEmptyEmailPassword() {
    // GIVEN: User is on the login page
    LoginPage loginPage = new LoginPage(driver);

    // WHEN: User tries to login without providing email and password
    loginPage.login();

    // THEN: User should remain on the login page
    String expectedURL = "https://qa.koel.app/";
    Assert.assertEquals(driver.getCurrentUrl(), expectedURL);
  }

  @Test
  public void loginValidEmailPassword() {
    // GIVEN: User is on the login page
    LoginPage loginPage = new LoginPage(driver);
    HomePage homePage = new HomePage(driver);

    // WHEN: User provides valid email and password and clicks submit
    loginPage.provideEmail("demo@testpro.io").providePassword("te$t$tudent")
        .clickSubmit();

    // THEN: User should be logged in and see the user avatar icon on the home page
    Assert.assertTrue(homePage.getUserAvatarIcon().isDisplayed());
  }

  @Test
  public void loginInvalidEmailValidPassword() {
    // GIVEN: User is on the login page
    LoginPage loginPage = new LoginPage(driver);

    // WHEN: User provides an invalid email and a valid password and clicks submit
    loginPage.login();
    loginPage.provideEmail("christina.taylor@test.io");
    loginPage.providePassword("te$t$tudent");

    // THEN: User should remain on the login page
    String expectedURL = "https://qa.koel.app/";
    Assert.assertEquals(driver.getCurrentUrl(), expectedURL);
  }

  @Test
  public void loginEmailEmptyPassword() {
    // GIVEN: User is on the login page
    LoginPage loginPage = new LoginPage(driver);

    // WHEN: User provides an email but no password and clicks submit
    loginPage.provideEmail("demo@testpro.io");
    loginPage.providePassword(" ");

    // THEN: User should remain on the login page
    String expectedURL = "https://qa.koel.app/";
    Assert.assertEquals(driver.getCurrentUrl(), expectedURL);
  }

  @Test(dataProvider = "NegativeLoginTestData", dataProviderClass = TestDataProvider.class)
  public void loginNegativeTest(String email, String password)  {
    // GIVEN: User is on the login page
    LoginPage loginPage = new LoginPage(driver);

    // WHEN: User tries to login with invalid credentials provided by the data provider
    loginPage.login();

    // THEN: User should remain on the login page
    String expectedURL = "https://qa.koel.app/";
    Assert.assertEquals(driver.getCurrentUrl(), expectedURL);
  }
}
```